server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -121,28 +207,189 @@ public ClusterState closeIndices(ClusterState currentState, final Index[] indice
+                                                       final ActionListener<ReplicationResponse> listener) {
+            final ShardId shardId = shardRoutingTable.shardId();
+            if (shardRoutingTable.primaryShard().unassigned()) {
+                logger.debug("primary shard {} is unassigned, ignoring", shardId);
+                final ReplicationResponse response = new ReplicationResponse();
+                response.setShardInfo(new ReplicationResponse.ShardInfo(shardRoutingTable.size(), shardRoutingTable.size()));
+                listener.onResponse(response);
+                return;
+            }
+            final TransportVerifyShardBeforeCloseAction.ShardRequest shardRequest =
+                new TransportVerifyShardBeforeCloseAction.ShardRequest(shardId);
+            if (timeout != null) {
+                shardRequest.timeout(timeout);
+            }
+            // TODO propagate a task id from the parent CloseIndexRequest to the ShardCloseRequests
+            transportVerifyShardBeforeCloseAction.execute(shardRequest, listener);
+        }
+    }
+
+    /**
+     * Step 3 - Move index states from OPEN to CLOSE in cluster state for indices that are ready for closing.
+     */
+    static ClusterState closeRoutingTable(final ClusterState currentState, final Map<Index, AcknowledgedResponse> results) {
+        final MetaData.Builder metadata = MetaData.builder(currentState.metaData());
+        final ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
+        final RoutingTable.Builder routingTable = RoutingTable.builder(currentState.routingTable());
+
+        final Set<String> closedIndices = new HashSet<>();
+        for (Map.Entry<Index, AcknowledgedResponse> result : results.entrySet()) {
+            final Index index = result.getKey();
+            try {
+                final IndexMetaData indexMetaData = metadata.getSafe(index);
+                if (indexMetaData.getState() != IndexMetaData.State.CLOSE) {
+                    if (result.getValue().isAcknowledged()) {
+                        logger.debug("closing index {} succeed, removing index routing table", index);
+                        metadata.put(IndexMetaData.builder(indexMetaData).state(IndexMetaData.State.CLOSE));
+                        routingTable.remove(index.getName());
+                        closedIndices.add(index.getName());
+                    } else {
+                        logger.debug("closing index {} failed, removing index block because: {}", index, result.getValue());
+                        blocks.removeIndexBlock(index.getName(), INDEX_CLOSED_BLOCK);
+                    }
+                }
+            } catch (final IndexNotFoundException e) {
+                logger.debug("index {} has been deleted since it was blocked before closing, ignoring", index);
+            }
+        }
+        logger.debug("closing indices {}", closedIndices);
```

Comment:
did you mean to log this at info level? Should we log two messages, one when closing is initiated and one when it is completed? This could then read as "completed closing of indices {}".

Reply:
Yes, it should be INFO and 2 messages make sense.